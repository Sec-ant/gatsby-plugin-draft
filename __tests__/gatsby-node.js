const mockdate = require('mockdate');
const moment = require('moment-timezone');
const { onCreateNode } = require('../gatsby-node');

test('not MarkdownRemark type node', () => {
  const createNodeField = jest.fn();
  const actions = { createNodeField };

  const node = {
    internal: {
      type: 'NotMarkdownRemark',
    },
  };

  onCreateNode({
    node,
    actions,
  });

  expect(createNodeField).toHaveBeenCalledTimes(0);
});

test('MarkdownRemark without date', () => {
  const createNodeField = jest.fn();
  const actions = { createNodeField };

  const node = {
    internal: {
      type: 'MarkdownRemark',
    },
    frontmatter: {
      title: 'post title',
    },
  };

  onCreateNode({
    node,
    actions,
  });

  expect(createNodeField).toMatchSnapshot();
  expect(createNodeField.mock.calls[0][0].name).toBe('draft');
  expect(createNodeField.mock.calls[0][0].value).toBeFalsy();
  expect(createNodeField).toHaveBeenCalledTimes(1);
});

test('draft should be false, (Same or Before)', () => {
  mockdate.set(moment.tz('2019-02-05', 'UTC'));

  const createNodeField = jest.fn();
  const actions = { createNodeField };

  const node = {
    internal: {
      type: 'MarkdownRemark',
    },
    frontmatter: {
      title: 'post title',
      date: '2019-02-07',
    },
  };

  onCreateNode({
    node,
    actions,
  });

  expect(createNodeField).toMatchSnapshot();
  expect(createNodeField.mock.calls[0][0].name).toBe('draft');
  expect(createNodeField.mock.calls[0][0].value).toBeFalsy();
  expect(createNodeField).toHaveBeenCalledTimes(1);

  mockdate.reset();
});

test('draft should be false, (not production)', () => {
  mockdate.set(moment.tz('2019-02-09', 'UTC'));

  const createNodeField = jest.fn();
  const actions = { createNodeField };

  const node = {
    internal: {
      type: 'MarkdownRemark',
    },
    frontmatter: {
      title: 'post title',
      date: '2019-02-07',
    },
  };

  onCreateNode({
    node,
    actions,
  });

  expect(createNodeField).toMatchSnapshot();
  expect(createNodeField.mock.calls[0][0].name).toBe('draft');
  expect(createNodeField.mock.calls[0][0].value).toBeFalsy();
  expect(createNodeField).toHaveBeenCalledTimes(1);

  mockdate.reset();
});

test('draft should be true', () => {
  mockdate.set(moment.tz('2019-02-05', 'UTC'));
  const env = process.env.NODE_ENV;
  process.env.NODE_ENV = 'production';

  const createNodeField = jest.fn();
  const actions = { createNodeField };

  const node = {
    internal: {
      type: 'MarkdownRemark',
    },
    frontmatter: {
      title: 'post title',
      date: '2019-02-07',
    },
  };

  onCreateNode({
    node,
    actions,
  });

  expect(createNodeField).toMatchSnapshot();
  expect(createNodeField.mock.calls[0][0].name).toBe('draft');
  expect(createNodeField.mock.calls[0][0].value).toBeTruthy();
  expect(createNodeField).toHaveBeenCalledTimes(1);

  process.env.NODE_ENV = env;
  mockdate.reset();
});

test('draft should be false (timezone)', () => {
  mockdate.set(moment.tz('2019-02-05', 'UTC'));
  const env = process.env.NODE_ENV;
  process.env.NODE_ENV = 'production';

  const createNodeField = jest.fn();
  const actions = { createNodeField };

  const node = {
    internal: {
      type: 'MarkdownRemark',
    },
    frontmatter: {
      title: 'post title',
      date: '2019-02-05',
    },
  };

  onCreateNode(
    {
      node,
      actions,
    },
    {
      timezone: 'Asia/Tokyo',
    }
  );

  expect(createNodeField).toMatchSnapshot();
  expect(createNodeField.mock.calls[0][0].name).toBe('draft');
  expect(createNodeField.mock.calls[0][0].value).toBeFalsy();
  expect(createNodeField).toHaveBeenCalledTimes(1);

  process.env.NODE_ENV = env;
  mockdate.reset();
});

test('draft should be true (timezone)', () => {
  mockdate.set(moment.tz('2019-02-05', 'UTC'));
  const env = process.env.NODE_ENV;
  process.env.NODE_ENV = 'production';

  const createNodeField = jest.fn();
  const actions = { createNodeField };

  const node = {
    internal: {
      type: 'MarkdownRemark',
    },
    frontmatter: {
      title: 'post title',
      date: '2019-02-05 12:00',
    },
  };

  onCreateNode(
    {
      node,
      actions,
    },
    {
      timezone: 'Asia/Tokyo',
    }
  );

  expect(createNodeField).toMatchSnapshot();
  expect(createNodeField.mock.calls[0][0].name).toBe('draft');
  expect(createNodeField.mock.calls[0][0].value).toBeTruthy();
  expect(createNodeField).toHaveBeenCalledTimes(1);

  process.env.NODE_ENV = env;
  mockdate.reset();
});

test('draft set to true should be respected if enabled', () => {
  const createNodeField = jest.fn();
  const actions = { createNodeField };

  const node = {
    internal: {
      type: 'MarkdownRemark',
    },
    frontmatter: {
      title: 'post title',
      draft: true,
    },
  };

  const pluginOptions = { respectExplicitDraft: true };

  onCreateNode(
    {
      node,
      actions,
    },
    pluginOptions
  );

  expect(createNodeField).toMatchSnapshot();
  expect(createNodeField.mock.calls[0][0].name).toBe('draft');
  expect(createNodeField.mock.calls[0][0].value).toBeTruthy();
  expect(createNodeField).toHaveBeenCalledTimes(1);
});

test('draft set to true should be ignored for default options', () => {
  const createNodeField = jest.fn();
  const actions = { createNodeField };

  const node = {
    internal: {
      type: 'MarkdownRemark',
    },
    frontmatter: {
      title: 'post title',
      draft: true,
    },
  };

  onCreateNode({
    node,
    actions,
  });

  expect(createNodeField).toMatchSnapshot();
  expect(createNodeField.mock.calls[0][0].name).toBe('draft');
  expect(createNodeField.mock.calls[0][0].value).toBeFalsy();
  expect(createNodeField).toHaveBeenCalledTimes(1);
});

test('draft should be true, if explicitly set, even if the date is past', () => {
  mockdate.set(moment.tz('2019-02-23', 'UTC'));
  const env = process.env.NODE_ENV;
  process.env.NODE_ENV = 'production';

  const createNodeField = jest.fn();
  const actions = { createNodeField };

  const node = {
    internal: {
      type: 'MarkdownRemark',
    },
    frontmatter: {
      title: 'post title',
      date: '2019-02-07',
      draft: true,
    },
  };

  const pluginOptions = { respectExplicitDraft: true };

  onCreateNode(
    {
      node,
      actions,
    },
    pluginOptions
  );

  expect(createNodeField).toMatchSnapshot();
  expect(createNodeField.mock.calls[0][0].name).toBe('draft');
  expect(createNodeField.mock.calls[0][0].value).toBeTruthy();
  expect(createNodeField).toHaveBeenCalledTimes(1);

  mockdate.reset();
});
