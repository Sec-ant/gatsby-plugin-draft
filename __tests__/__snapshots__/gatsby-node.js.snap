// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkdownRemark without date 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "name": "draft",
        "node": Object {
          "frontmatter": Object {
            "title": "post title",
          },
          "internal": Object {
            "type": "MarkdownRemark",
          },
        },
        "value": false,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`draft set to true should be ignored for default options 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "name": "draft",
        "node": Object {
          "frontmatter": Object {
            "draft": true,
            "title": "post title",
          },
          "internal": Object {
            "type": "MarkdownRemark",
          },
        },
        "value": false,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`draft set to true should be respected if enabled 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "name": "draft",
        "node": Object {
          "frontmatter": Object {
            "draft": true,
            "title": "post title",
          },
          "internal": Object {
            "type": "MarkdownRemark",
          },
        },
        "value": true,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`draft should be false (timezone) 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "name": "draft",
        "node": Object {
          "frontmatter": Object {
            "date": "2019-02-05",
            "title": "post title",
          },
          "internal": Object {
            "type": "MarkdownRemark",
          },
        },
        "value": false,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`draft should be false, (Same or Before) 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "name": "draft",
        "node": Object {
          "frontmatter": Object {
            "date": "2019-02-07",
            "title": "post title",
          },
          "internal": Object {
            "type": "MarkdownRemark",
          },
        },
        "value": false,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`draft should be false, (not production) 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "name": "draft",
        "node": Object {
          "frontmatter": Object {
            "date": "2019-02-07",
            "title": "post title",
          },
          "internal": Object {
            "type": "MarkdownRemark",
          },
        },
        "value": false,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`draft should be true (timezone) 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "name": "draft",
        "node": Object {
          "frontmatter": Object {
            "date": "2019-02-05 12:00",
            "title": "post title",
          },
          "internal": Object {
            "type": "MarkdownRemark",
          },
        },
        "value": true,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`draft should be true 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "name": "draft",
        "node": Object {
          "frontmatter": Object {
            "date": "2019-02-07",
            "title": "post title",
          },
          "internal": Object {
            "type": "MarkdownRemark",
          },
        },
        "value": true,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`draft should be true, if explicitly set, even if the date is past 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "name": "draft",
        "node": Object {
          "frontmatter": Object {
            "date": "2019-02-07",
            "draft": true,
            "title": "post title",
          },
          "internal": Object {
            "type": "MarkdownRemark",
          },
        },
        "value": true,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
